modules/spring-config/src/main/java/org/mule/runtime/config/internal/resolvers/DependencyGraphBeanDependencyResolver.java

```diff
@@ -0,0 +1,212 @@
+    Object currentObject = springRegistry.get(beanName);
+    final DependencyNode currentNode = new DependencyNode(currentObject);
+
+    addDirectDependency(beanName, currentObject, currentNode, new HashSet<>());
+
+    return currentNode.getChildren()
+        .stream()
+        .map(DependencyNode::getObject)
+        .collect(toList());
+  }
+
+  @Override
+  public ConfigurationDependencyResolver getConfigurationDependencyResolver() {
+    return configurationDependencyResolver;
+  }
+
+  /**
+   * Provides only direct dependencies/required components for the object provided
+   * 
+   * @return direct children(required components) of the current object If the component is already visited for the current
+   *         bucket, we don't need to get dependencies again, return emptyList If the component is already visited for any other
+   *         buckets, but not for this bucket, return the dependencies already saved If it was never visited before, get direct
+   *         dependencies for the object
+   */
+  public List<BeanWrapper> getDirectBeanDependencies(BeanWrapper currentObject, int bucketIndex) {
+    // If the component was already processed for the current bucket(graph), it won't process it
+    visitedComponentsForBuckets.putIfAbsent(bucketIndex, new HashSet<>());
+    if (visitedComponentsForBuckets.get(bucketIndex).stream().map(x -> x.getWrappedObject())
+        .anyMatch(x -> x.equals(currentObject.getWrappedObject()))) {
+      return emptyList();
+    }
+    visitedComponentsForBuckets.get(bucketIndex).add(currentObject);
+
+    // If the component was already processed for any other buckets,
+    // it returns the direct dependencies that were saved before to build a dependency graph for the current bucket
+    if (visitedComponents.containsKey(currentObject)) {
+      return visitedComponents.get(currentObject);
+    }
+
+    // Find direct dependencies for current
+    final DependencyNode currentNode = new DependencyNode(currentObject.getName(), currentObject.getWrappedObject());
+    addDirectDependency(currentObject.getName(), currentObject.getWrappedObject(), currentNode, new HashSet<>());
+
+    List<BeanWrapper> directDependencies = currentNode.getChildren()
+        .stream()
+        .map(DependencyNode::getNameAndObject)
+        .collect(toList());
+
+    // save the dependencies of th current component for future usages
+    visitedComponents.put(currentObject, directDependencies);
+
+    return directDependencies;
+  }
+
+  /**
+   * Provides all the dependencies/required components for the object provided
+   *
+   * @param beanName    the name of the bean to resolve dependencies
+   * @param bucketIndex the bucket(graph) we calculate the dependencies for
+   * @return a map with any relevant object as key, direct dependencies of the key as value
+   *
+   */
+  public Map<BeanWrapper, List<BeanWrapper>> getTransitiveDependencies(String beanName, int bucketIndex) {
```

Comment:
It is called one per object that fits the bucket. If the object has already been processed inside this bucket, then we don't need to calculate its dependencies so we return an empty list since all nodes and edges that could come from calculating should already be present in the graph.

Reply:
Yes, you are right I forgot `getDirectBeanDependencies` would return an empty list if already processed. However if this happens very often for the same object (I mean, receiving an already processed object), maybe we could check that here too and avoid the creation of all these helper objects (mainly `processedComponents` and `queue`). But I wouldn't do it unless we can confirm that this is indeed called many times.

Have you considered the case in which it was processed but for a different bucket? in that case we would be actually rebuilding a result that could be cached, right? But maybe the memory footprint could be a concern in that case, unless we use a bounded cache. Again, I wouldn't change anything unless it can be confirmed that it makes a difference.

Reply:
Yes, the duplicate check/caching is happening when calculating direct dependencies only at this moment. I will put caching the result of getTransitiveDependencies as subtask of benchmark ticket. Since we are going through the objects from the lookup and lookup doesn't have much repetition, getTransitiveDependencies won't be called many times for the same object.

Reply:
Works for me.